text/0250-Composite_Samplers.md

```diff
@@ -0,0 +1,204 @@
+Otherwise, if any of the delegate Decisions is `RECORD_ONLY`, the composite sampler MUST return `RECORD_ONLY` Decision.
+- The set of span Attributes to be added to the `Span` is the sum of the sets of Attributes as provided by the delegate samplers within their `SamplingResults`s.
+- The `TraceState` to be used with the new `Span` is obtained by cumulatively applying all the potential modfications of the parent `TraceState` by the delegate samplers, with special handling of the `th` sub-key (the sampling rejection `THRESHOLD`) for the `ot` entry as described below.
+
+If the final sampling Decision is `DROP` or `RECORD_ONLY`, the `th` entry MUST be removed.
+If the sampling Decision is `RECORD_AND_SAMPLE`, and there's no `th` entry in any of the `TraceState` provided by the delegates that decided to `RECORD_AND_SAMPLE`, the `th` entry MUST be also removed.
+Otherwise, the resulting `TraceState` MUST contain `th` entry with the `THRESHOLD` value being the maximum of all the `THRESHOLD` values as reported by those delegates that decided to `RECORD_AND_SAMPLE`.
+
+Each delegate sampler MUST be given a chance to participate in the sampling decision as described above and MUST see the same _parent_ state. The order of the delegate samplers does not matter, as long as there's no overlap in the Attribute Keys or the trace state keys (other than `th`) that they use.
+
+### Conjunction
+
+`Conjunction` is a composite sampler which takes two Samplers (delegates) as the arguments. These delegate samplers will be hereby referenced as First and Second.
+
+Upon invocation of its `shouldSample` method, the Conjunction sampler MUST invoke `shouldSample` method on the First sampler, passing the same arguments as received, and examine the received sampling Decision. Upon receiving `DROP` or `RECORD_ONLY` decision it MUST return the SamplingResult from the First sampler, and it MUST NOT proceed with querying the Second sampler. If the sampling decision from the First sampler is `RECORD_AND_SAMPLE`, the Conjunction sampler MUST invoke `shouldSample` method on the Second sampler.
+If the sampling Decision from the Second sampler is `RECORD_AND_SAMPLE`, the Conjunction sampler MUST return a `SamplingResult` which is constructed as follows:
+
+- The sampling Decision is `RECORD_AND_SAMPLE`.
+- The set of span Attributes to be added to the `Span` is the sum of the sets of Attributes as provided by the First and the Second sampler.
+- The `TraceState` to be used with the new `Span` is obtained by cumulatively applying the potential modfications from the First and Second sampler, with special handling of the `th` sub-key (the sampling rejection `THRESHOLD`) for the `ot` entry as described below.
+
+If both First and Second samplers provided `th` entry in the returned `TraceState`, the resulting `TraceState` MUST contain `th` entry with the `THRESHOLD` being maximum of the `THRESHOLD`s provided by the First and Second samplers. Otherwise, the `th` entry MUST be removed.
```

Comment:
Doesn't the composite sampler break consistent sampling? The threshold value must always be chosen independently of the `R`-value. However, the use of the threshold value of the second sampler depends on the decision of the first sampler, whose decision on sampling depends on `T <= R` and thus also on `R`. Therefore, there is an indirect dependency on `R` which is not allowed.

Reply:
Yes, you are right. It looks like the `Conjunction` sampler cannot be made compatible with consistent probability sampling. However, it is still needed, for example if someone want sto use a reservoir-based rate limiting sampler as the Second sampler.

Reply:
When composing samplers, the composition must be based on the thresholds and must happen before the actual sampling decision, which compares the threshold with the random value. In other words, referring to the Java prototypes, there must be only a single `shouldSample` invovation for every span (https://github.com/open-telemetry/opentelemetry-java-contrib/blob/6a44e04a76bbe6b9844555f9b37e777c2a23103f/consistent-sampling/src/main/java/io/opentelemetry/contrib/sampler/consistent/ConsistentSampler.java#L232).

However, the sampler may delegate during the threshold computation to other samplers like it is done, for example, in the parent-based sampler (https://github.com/open-telemetry/opentelemetry-java-contrib/blob/6a44e04a76bbe6b9844555f9b37e777c2a23103f/consistent-sampling/src/main/java/io/opentelemetry/contrib/sampler/consistent56/ConsistentParentBasedSampler.java#L40).

To cover the example above, it should be possible to extend the rate-limiting consistent sampler and to provide a factory method such as (compare https://github.com/open-telemetry/opentelemetry-java-contrib/blob/6a44e04a76bbe6b9844555f9b37e777c2a23103f/consistent-sampling/src/main/java/io/opentelemetry/contrib/sampler/consistent/ConsistentSampler.java#L117)
```
ConsistentSampler rateLimited(double targetSpansPerSecondLimit, double adaptationTimeSeconds, ConsistentSampler delegateSampler);
```
that returns a rate-limiting sampler, that delegates the threshold computation to the `delegateSampler` and adapts the returned threshold in such a way to meet the rate limit. For that the rate-limiting sampler has to keep track of some statistics about the number of invovations and the thresholds returned by the delegate.

Reply:
I had considered adding a delegate sampler to the `RateLimitingConsistentSampler`, but opted out, because it addresses just one sampler, and, furthermore, I believe that when dealing with consistent probability sampling, `EachOf` handles this case correctly.
`Conjunction` is needed for reservoir based samplers.

Reply:
Added Approach Two, describing this idea.